batch/src/main/java/org/openmrs/analytics/JdbcFetchUtil.java

```diff
@@ -0,0 +1,205 @@
+package org.openmrs.analytics;
+
+import java.beans.PropertyVetoException;
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import com.google.gson.Gson;
+import com.mchange.v2.c3p0.ComboPooledDataSource;
+import org.apache.beam.sdk.coders.NullableCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.jdbc.JdbcIO;
+import org.apache.beam.sdk.transforms.*;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.openmrs.analytics.model.EventConfiguration;
+import org.openmrs.analytics.model.GeneralConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JdbcFetchUtil {
+	
+	private static final Logger log = LoggerFactory.getLogger(JdbcFetchUtil.class);
+	
+	private ComboPooledDataSource comboPooledDataSource;
+	
+	JdbcFetchUtil(String jdbcDriverClass, String jdbcUrl, String dbUser, String dbPassword, int dbcMaxPoolSize)
+	        throws IOException, SQLException, PropertyVetoException {
+		comboPooledDataSource = new ComboPooledDataSource();
+		comboPooledDataSource.setDriverClass(jdbcDriverClass);
+		comboPooledDataSource.setJdbcUrl(jdbcUrl);
+		comboPooledDataSource.setUser(dbUser);
+		comboPooledDataSource.setPassword(dbPassword);
+		comboPooledDataSource.setMaxPoolSize(dbcMaxPoolSize);
+		comboPooledDataSource.setInitialPoolSize(10); // TODO: make this configurable if issues arises
+	}
+	
+	public static class FetchUuIds extends PTransform<PCollection<KV<String, Integer>>, PCollection<String>> {
+		
+		private final String tableName;
+		
+		private final JdbcIO.DataSourceConfiguration dataSourceConfig;
+		
+		public FetchUuIds(String tableName, JdbcIO.DataSourceConfiguration dataSourceConfig) {
+			this.tableName = tableName;
+			this.dataSourceConfig = dataSourceConfig;
+		}
+		
+		@Override
+		public PCollection<String> expand(PCollection<KV<String, Integer>> idRanges) {
+			String tableId = this.tableName + "_id";
+			return idRanges.apply(String.format("Read UUIDs from %s", this.tableName),
+			    JdbcIO.<KV<String, Integer>, String> readAll().withDataSourceConfiguration(this.dataSourceConfig)
+			            .withParameterSetter(new JdbcIO.PreparedStatementSetter<KV<String, Integer>>() {
+				            
+				            @Override
+				            public void setParameters(KV<String, Integer> element, PreparedStatement preparedStatement)
+				                    throws Exception {
+					            String[] range = Objects.requireNonNull(element.getKey()).split(",");
```

Comment:
Are you using the values in these `KV`s anywhere? If not, can we use a `Map<Integer, Integer>` as the returned value of `createIdRanges` (returning the two integers) and avoid the string concatenation (there) and split (here) altogether?

Reply:
In the initial implementation, I was using the Key. For this new approach, I don't think we are using it anywhere! Thanks for the suggestion, changing to  `Map<Integer, Integer> `